File: order-service/src/main/java/com/catius/order/domain/Order.java
package com.catius.order.domain;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;


@Entity
@Table(name = "orders")
@Getter
@NoArgsConstructor
@Builder
@AllArgsConstructor
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private Long customerId;

    private Long productId;

    private int quantity;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderStatus status;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    public static Order create(Long customerId, Long productId, int quantity) {
        return Order.builder()
                .customerId(customerId)
                .productId(productId)
                .quantity(quantity)
                .status(OrderStatus.PENDING)
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();
    }

    public void confirm() {
        this.status = OrderStatus.CONFIRMED;
        this.updatedAt = LocalDateTime.now();
    }

    public void cancel() {
        this.status = OrderStatus.CANCELLED;
        this.updatedAt = LocalDateTime.now();
    }

}

File: order-service/src/main/kotlin/com/catius/order/domain/Order.kt
package com.catius.order.domain

import java.time.Instant

enum class OrderStatus { PENDING, CONFIRMED, FAILED }

data class OrderItem(
    val productId: Long,
    val quantity: Int,
) {
    init { require(quantity > 0) { "수량은 1 이상이어야 합니다: $quantity" } }
}

data class Order(
    val id: Long?,
    val customerId: Long,
    val items: List<OrderItem>,
    val status: OrderStatus,
    val createdAt: Instant,
    val updatedAt: Instant,
) {
    init {
        require(items.isNotEmpty()) { "주문 항목이 비어 있습니다" }
        require(items.distinctBy { it.productId }.size == items.size) {
            "동일 상품이 중복으로 포함될 수 없습니다"
        }
    }

    val totalQuantity: Int get() = items.sumOf { it.quantity }

    fun confirm(at: Instant): Order {
        check(status == OrderStatus.PENDING) { "PENDING 에서만 확정 가능: 현재=$status" }
        return copy(status = OrderStatus.CONFIRMED, updatedAt = at)
    }

    fun fail(at: Instant): Order {
        check(status == OrderStatus.PENDING) { "PENDING 에서만 실패 처리 가능: 현재=$status" }
        return copy(status = OrderStatus.FAILED, updatedAt = at)
    }

    companion object {
        fun create(customerId: Long, items: List<OrderItem>, now: Instant) = Order(
            id = null,
            customerId = customerId,
            items = items,
            status = OrderStatus.PENDING,
            createdAt = now,
            updatedAt = now,
        )
    }
}
